File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/Case.scala
package org.enso.compiler.core.ir
package expression

import org.enso.compiler.core.Implicits.{ShowPassData, ToStringHelper}
import org.enso.compiler.core.{IR, Identifier}
import org.enso.compiler.core.IR.{indentLevel, mkIndent}

import java.util.UUID

/** The Enso case expression. */
sealed trait Case extends Expression {

  /** @inheritdoc */
  override def mapExpressions(
    fn: java.util.function.Function[Expression, Expression]
  ): Case

  /** @inheritdoc */
  override def setLocation(location: Option[IdentifiedLocation]): Case

  /** @inheritdoc */
  override def duplicate(
    keepLocations: Boolean   = true,
    keepMetadata: Boolean    = true,
    keepDiagnostics: Boolean = true,
    keepIdentifiers: Boolean = false
  ): Case
}

object Case {

  /** The main body of the Enso case expression.
    *
    * @param scrutinee   the expression whose value is being matched on
    * @param branches    the branches of the case expression
    * @param isNested    if true, the flag indicates that the expr represents a desugared nested case
    * @param location    the source location that the node corresponds to
    * @param passData    the pass metadata associated with this node
    * @param diagnostics compiler diagnostics for this node
    */
  sealed case class Expr(
    scrutinee: Expression,
    branches: Seq[Branch],
    isNested: Boolean,
    override val location: Option[IdentifiedLocation],
    override val passData: MetadataStorage      = new MetadataStorage(),
    override val diagnostics: DiagnosticStorage = DiagnosticStorage()
  ) extends Case
      with IRKind.Primitive
      with LazyId {

    def this(
      scrutinee: Expression,
      branches: Seq[Branch],
      location: Option[IdentifiedLocation],
      passData: MetadataStorage,
      diagnostics: DiagnosticStorage
    ) = {
      this(scrutinee, branches, false, location, passData, diagnostics)
    }

    /** Creates a copy of `this`.
      *
      * @param scrutinee   the expression whose value is being matched on
      * @param branches    the branches of the case expression
      * @param isNested    if true, the flag indicates that the expr represents a desugared nested case
      * @param location    the source location that the node corresponds to
      * @param passData    the pass metadata associated with this node
      * @param diagnostics compiler diagnostics for this node
      * @param id          the identifier for the new node
      * @return a copy of `this`, updated with the specified values
      */
    def copy(
      scrutinee: Expression                = scrutinee,
      branches: Seq[Branch]                = branches,
      isNested: Boolean                    = isNested,
      location: Option[IdentifiedLocation] = location,
      passData: MetadataStorage            = passData,
      diagnostics: DiagnosticStorage       = diagnostics,
      id: UUID @Identifier                 = id
    ): Expr = {
      val res =
        Expr(scrutinee, branches, isNested, location, passData, diagnostics)
      res.id = id
      res
    }

    /** @inheritdoc */
    override def duplicate(
      keepLocations: Boolean   = true,
      keepMetadata: Boolean    = true,
      keepDiagnostics: Boolean = true,
      keepIdentifiers: Boolean = false
    ): Expr =
      copy(
        scrutinee = scrutinee.duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
        branches = branches.map(
          _.duplicate(
            keepLocations,
            keepMetadata,
            keepDiagnostics,
            keepIdentifiers
          )
        ),
        isNested = isNested,
        location = if (keepLocations) location else None,
        passData =
          if (keepMetadata) passData.duplicate else new MetadataStorage(),
        diagnostics =
          if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
        id = if (keepIdentifiers) id else null
      )

    /** @inheritdoc */
    override def setLocation(location: Option[IdentifiedLocation]): Expr =
      copy(location = location)

    /** @inheritdoc */
    override def mapExpressions(
      fn: java.util.function.Function[Expression, Expression]
    ): Expr = {
      copy(
        scrutinee = fn(scrutinee),
        branches.map(_.mapExpressions(fn))
      )
    }

    /** @inheritdoc */
    override def toString: String =
      s"""
         |Case.Expr(
         |scrutinee = $scrutinee,
         |branches = $branches,
         |isNested = $isNested,
         |location = $location,
         |passData = ${this.showPassData},
         |diagnostics = $diagnostics,
         |id = $id
         |)
         |""".toSingleLine

    /** @inheritdoc */
    override def children: List[IR] = scrutinee :: branches.toList

    /** @inheritdoc */
    override def showCode(indent: Int): String = {
      val newIndent = indent + indentLevel
      val headerStr = s"case ${scrutinee.showCode(indent)} of"
      val branchesStr = branches
        .map(mkIndent(newIndent) + _.showCode(newIndent))
        .mkString("\n")

      s"$headerStr\n$branchesStr"
    }
  }

  object Expr {
    def apply(
      scrutinee: Expression,
      branches: Seq[Branch],
      location: Option[IdentifiedLocation]
    ): Expr =
      apply(
        scrutinee,
        branches,
        location,
        new MetadataStorage(),
        new DiagnosticStorage()
      )

    def apply(
      scrutinee: Expression,
      branches: Seq[Branch],
      location: Option[IdentifiedLocation],
      passData: MetadataStorage,
      diagnostics: DiagnosticStorage
    ): Expr = new Expr(scrutinee, branches, location, passData, diagnostics)
  }

  /** A branch in a case statement.
    *
    * @param pattern        the pattern that attempts to match against the scrutinee
    * @param expression     the expression that is executed if the pattern matches
    * @param terminalBranch the flag indicating whether the branch represents the final pattern to be checked
    * @param location       the source location that the node corresponds to
    * @param passData       the pass metadata associated with this node
    * @param diagnostics    compiler diagnostics for this node
    */
  sealed case class Branch(
    pattern: Pattern,
    expression: Expression,
    terminalBranch: Boolean,
    override val location: Option[IdentifiedLocation],
    override val passData: MetadataStorage      = new MetadataStorage(),
    override val diagnostics: DiagnosticStorage = DiagnosticStorage()
  ) extends Case
      with IRKind.Primitive
      with LazyId {

    def this(
      pattern: Pattern,
      expression: Expression,
      location: Option[IdentifiedLocation],
      passData: MetadataStorage,
      diagnostics: DiagnosticStorage
    ) = {
      this(pattern, expression, true, location, passData, diagnostics)
    }

    /** Creates a copy of `this`.
      *
      * @param pattern     the pattern that attempts to match against the scrutinee
      * @param expression  the expression that is executed if the pattern matches
      * @param location    the source location that the node corresponds to
      * @param passData    the pass metadata associated with this node
      * @param diagnostics compiler diagnostics for this node
      * @param id          the identifier for the new node
      * @return a copy of `this`, updated with the specified values
      */
    def copy(
      pattern: Pattern                     = pattern,
      expression: Expression               = expression,
      terminalBranch: Boolean              = terminalBranch,
      location: Option[IdentifiedLocation] = location,
      passData: MetadataStorage            = passData,
      diagnostics: DiagnosticStorage       = diagnostics,
      id: UUID @Identifier                 = id
    ): Branch = {
      val res = Branch(
        pattern,
        expression,
        terminalBranch,
        location,
        passData,
        diagnostics
      )
      res.id = id
      res
    }

    /** @inheritdoc */
    override def duplicate(
      keepLocations: Boolean   = true,
      keepMetadata: Boolean    = true,
      keepDiagnostics: Boolean = true,
      keepIdentifiers: Boolean = false
    ): Branch =
      copy(
        pattern = pattern.duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
        expression = expression.duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
        terminalBranch = terminalBranch,
        location       = if (keepLocations) location else None,
        passData =
          if (keepMetadata) passData.duplicate else new MetadataStorage(),
        diagnostics =
          if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
        id = if (keepIdentifiers) id else null
      )

    /** @inheritdoc */
    override def setLocation(location: Option[IdentifiedLocation]): Branch =
      copy(location = location)

    /** @inheritdoc */
    override def mapExpressions(
      fn: java.util.function.Function[Expression, Expression]
    ): Branch = {
      copy(pattern = pattern.mapExpressions(fn), expression = fn(expression))
    }

    /** @inheritdoc */
    override def toString: String =
      s"""
         |Case.Branch(
         |pattern = $pattern,
         |expression = $expression,
         |terminalBranch = $terminalBranch,
         |location = $location,
         |passData = ${this.showPassData},
         |diagnostics = $diagnostics,
         |id = $id
         |)
         |""".toSingleLine

    /** @inheritdoc */
    override def children: List[IR] = List(pattern, expression)

    /** @inheritdoc */
    override def showCode(indent: Int): String = {
      val newIndent = indent + indentLevel
      val bodyStr = if (expression.isInstanceOf[Expression.Block]) {
        s"\n${mkIndent(newIndent)}${expression.showCode(newIndent)}"
      } else {
        s"${expression.showCode(indent)}"
      }
      s"${pattern.showCode(indent)} -> $bodyStr"
    }
  }

  object Branch {
    def apply(
      pattern: Pattern,
      expression: Expression,
      location: Option[IdentifiedLocation]
    ): Branch =
      apply(
        pattern,
        expression,
        location,
        new MetadataStorage(),
        new DiagnosticStorage()
      )

    def apply(
      pattern: Pattern,
      expression: Expression,
      location: Option[IdentifiedLocation],
      passData: MetadataStorage,
      diagnostics: DiagnosticStorage
    ): Branch =
      new Branch(pattern, expression, location, passData, diagnostics)
  }
}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/expression/Case.java
package org.enso.compiler.core.ir.expression;

import java.util.UUID;
import java.util.function.Function;
import org.enso.compiler.core.IR;
import org.enso.compiler.core.ir.DiagnosticStorage;
import org.enso.compiler.core.ir.Expression;
import org.enso.compiler.core.ir.IRKind;
import org.enso.compiler.core.ir.IdentifiedLocation;
import org.enso.compiler.core.ir.MetadataStorage;
import org.enso.compiler.core.ir.Pattern;
import org.enso.runtime.parser.dsl.GenerateFields;
import org.enso.runtime.parser.dsl.GenerateIR;
import org.enso.runtime.parser.dsl.IRChild;
import org.enso.runtime.parser.dsl.IRField;
import scala.Option;
import scala.collection.immutable.List;

public interface Case extends Expression {

  @Override
  Case mapExpressions(Function<Expression, Expression> fn);

  @Override
  Case setLocation(Option<IdentifiedLocation> location);

  @Override
  Case duplicate(
      boolean keepLocations,
      boolean keepMetadata,
      boolean keepDiagnostics,
      boolean keepIdentifiers);

  @GenerateIR(interfaces = {Case.class, IRKind.Primitive.class})
  final class Expr extends CaseExprGen {
    public static Builder builder() {
      return new Builder();
    }

    @GenerateFields
    public Expr(
        @IRChild Expression scrutinee,
        @IRChild List<Branch> branches,
        @IRField boolean isNested,
        IdentifiedLocation identifiedLocation,
        MetadataStorage passData,
        DiagnosticStorage diagnostics) {
      super(scrutinee, branches, isNested, identifiedLocation, passData, diagnostics);
    }

    public Expr copy(Expression scrutinee, List<Branch> branches, boolean isNested) {
      return new Builder(this).scrutinee(scrutinee).branches(branches).isNested(isNested).build();
    }

    public Expr copy(Expression scrutinee, List<Branch> branches) {
      return new Builder(this).scrutinee(scrutinee).branches(branches).build();
    }

    public Expr copy(List<Branch> branches) {
      return new Builder(this).branches(branches).build();
    }

    @Override
    public String showCode(int indent) {
      var newIndent = indent + indentLevel;
      var headerStr = "case " + scrutinee().showCode(indent) + " of";
      var branchesStr =
          branches().map(br -> IR.mkIndent(newIndent) + br.showCode(newIndent)).mkString("\n");
      return headerStr + "\n" + branchesStr;
    }
  }

  @GenerateIR(interfaces = {Case.class, IRKind.Primitive.class})
  final class Branch extends CaseBranchGen {
    public static Builder builder() {
      return new Builder();
    }

    @GenerateFields
    public Branch(
        @IRChild Pattern pattern,
        @IRChild Expression expression,
        @IRField boolean terminalBranch,
        IdentifiedLocation identifiedLocation,
        MetadataStorage passData,
        DiagnosticStorage diagnostics) {
      super(pattern, expression, terminalBranch, identifiedLocation, passData, diagnostics);
    }

    public Branch copy(Pattern pattern, Expression expression, boolean terminalBranch) {
      return new Builder(this)
          .pattern(pattern)
          .expression(expression)
          .terminalBranch(terminalBranch)
          .build();
    }

    public Branch copy(Expression expression, IdentifiedLocation identifiedLocation) {
      return new Builder(this).expression(expression).location(identifiedLocation).build();
    }

    public UUID id() {
      // TBD: this method should return null when there is no ID
      // but there is no way to check if the ID is null or not now
      // when the id field is private in a superclass
      return super.getId();
    }

    @Override
    public String showCode(int indent) {
      var newIndent = indent + indentLevel;
      String bodyStr;
      if (expression() instanceof Expression.Block block) {
        bodyStr = "\n" + IR.mkIndent(newIndent) + block.showCode(newIndent);
      } else {
        bodyStr = expression().showCode(indent);
      }
      return pattern().showCode(indent) + "->" + bodyStr;
    }
  }
}
